README.md
n-set
=======

[![Build Status](https://travis-ci.org/Daninet/n-set.svg?branch=master)](https://travis-ci.org/Daninet/n-set)
[![Coverage Status](https://coveralls.io/repos/github/Daninet/n-set/badge.svg?branch=master)](https://coveralls.io/github/Daninet/n-set?branch=master)
[![license](https://img.shields.io/github/license/Daninet/n-set.svg)](https://github.com/Daninet/n-set/blob/master/LICENSE)

Allows to put arrays into sets. Works similarly like ES6 Sets.

Contains type definitions for TypeScript.

Install
=======
```
npm i n-set
```

Example
=======
```javascript
const NSet = require('n-set');
const set = new NSet();
set.add(1);
set.add(2);
set.add(2);
console.log(set); // NSet { (1), (2) }
set.add(1, 2, 3);
set.add(1, 5);
set.add([1, 5]); // you can pass it as an array too
console.log(set); // NSet { (1), (1, 2, 3), (1, 5), (2) }
set.delete(1, 2, 3);
console.log(set); // NSet { (1), (1, 5), (2) }
console.log(set.has(1, 2, 3)); // false
console.log(set.has(1, 5)); // true
```
How it works?
=======
It uses ES6 Maps for storing each level of depth and an ES6 Set for the last one.

By example for storing (1, 2, 3) and (1, 2, 4), it will use internally this data structure:
```
Map( 1 => 
  Map ( 2 => 
    Set ( 3, 4 )
  )
)
```

Methods
=======
set.add(tuple)
-------
Appends the specified tuple of values to the end of Set. The values can be passed as an array or as parameters: `set.add(1, 2, 3)` has the same behaviour as `set.add([1, 2, 3])`. 

Note: Adding `undefined` value is not supported.

Returns the Set object.

set.clear()
-------
Removes all elements from a Set object.

set.delete(tuple)
-------
Removes the specified tuple from a Set object. The values can be passed as an array or as parameters: `set.delete(1, 2, 3)` has the same behaviour as `set.delete([1, 2, 3])`.
Returns `true` if an element in the Set object has been removed successfully; otherwise `false`.

set.entries()
-------
Returns a new `Iterator` object that contains an array of `[tuple, tuple]` for each element in the Set object.

set.forEach(func)
-------
Executes the provided function once for each value in the Set object.

set.has(tuple)
-------
Returns a boolean indicating whether an element with the specified tuple exists in a Set object or not.

set.size
-------
Returns the number of elements in the Set object.

set.values()
-------
Returns a new `Iterator` object that contains the values for each element in the Set object.

Performance
=======

Each test consists of three operations:
- .add() 250 000 tuples
- .has() on the 250 000 tuples 
- .delete() on the 250 000 tuples

1-depth has ~250 000 tuples of length 1. [v1] - Native ES6 Sets and Maps supports only these.

2-depth has ~250 000 tuples of length 2. [v1, v2]

3-depth has ~250 000 tuples of length 3. [v1, v2, v3]

4-depth has ~250 000 tuples of length 4. [v1, v2, v3, v4]

The source of this benchmark can be found in file `benchmark.js`.

```
Running it with Node.js v8.9.4 on Intel i7-7700K CPU.

After 5 runs:
------------
Map 1-depth  - Min: 23.359 ms, Max: 27.608 ms, Mean: 25.142 ms, Median: 23.847 ms
Set 1-depth  - Min: 22.383 ms, Max: 27.031 ms, Mean: 24.195 ms, Median: 24.385 ms
NSet 1-depth - Min: 41.212 ms, Max: 172.270 ms, Mean: 70.170 ms, Median: 41.787 ms
NSet 2-depth - Min: 97.894 ms, Max: 110.139 ms, Mean: 102.154 ms, Median: 110.139 ms
NSet 3-depth - Min: 206.468 ms, Max: 349.996 ms, Mean: 237.171 ms, Median: 208.478 ms
NSet 4-depth - Min: 972.355 ms, Max: 1079.711 ms, Mean: 1052.376 ms, Median: 1072.435 ms
```

License
=======
MIT


CHANGELOG.md
## 2.0.1 (July 31, 2019)

* Add keyword for TypeScript

## 2.0.0 (July 31, 2019)

* Migrated to TypeScript
* Added toJSON(), clone()
* Faster iteration over elements

## 1.0.1 (April 6, 2018)

* Add ESLint

## 1.0.0 (February 15, 2018)

* Initial release

test/Fuzz.test.js
'use strict';
/* global test, expect */

const NSet = require('../');
const NMAX = 50;

function randInt (min, max) {
  return Math.floor(Math.random() * (max - min + 1)) + min;
}

function compareArrayAndSet (arr, nset) {
  arr.forEach(it => {
    expect(nset.has(it)).toEqual(true);
    expect(nset.has([...it, 100])).toBe(false);
    expect(nset.has([...it.slice(0, -1), 51])).toBe(false);
  });
  const strElements = arr.map(f => JSON.stringify(f));
  nset.forEach(it => {
    const strArr = JSON.stringify(it);
    const res = strElements.find(f => f === strArr);
    expect(res).not.toBeUndefined();
  });
}

test('set-simple', () => {
  for (let n = 0; n < NMAX; n++) {
    const set = new Set();
    const nset = new NSet();
    for (let i = 0; i < 1000; i++) {
      const rnd = randInt(0, 100);
      expect(nset.has(rnd)).toBe(set.has(rnd));
      set.add(rnd);
      nset.add(rnd);
      expect(nset.has(rnd)).toBe(true);
    }
    expect(nset.size).toBe(set.size);
    for (let i = 0; i < 100; i++) {
      const rnd = randInt(0, 100);
      expect(nset.has(rnd)).toBe(set.has(rnd));
      expect(nset.delete(rnd)).toBe(set.delete(rnd));
      expect(nset.has(rnd)).toBe(false);
    }
    expect(nset.size).toBe(set.size);
    for (let i = 0; i <= 100; i++) {
      expect(nset.has(i)).toBe(set.has(i));
    }
  }
});

test('set-deep', () => {
  for (let n = 0; n < NMAX; n++) {
    let items = [];
    const nset = new NSet();
    for (let i = 0; i < 1000; i++) {
      const arr = [];
      const c = randInt(1, 9);
      for (let j = 0; j < c; j++) {
        arr.push(randInt(1, 50));
      }
      nset.add(arr);
      expect(nset.has(arr)).toBe(true);
      items.push(arr);
    }
    expect(nset.size).toBeLessThanOrEqual(1000);
    compareArrayAndSet(items, nset);
    const startSize = nset.size;
    for (let i = 0; i < 10; i++) {
      const arr = [];
      const c = randInt(1, 9);
      for (let j = 0; j < c; j++) {
        arr.push(randInt(1, 50));
      }
      if (nset.has(arr)) {
        continue;
      }
      expect(nset.delete(arr)).toBe(false);
    }
    expect(nset.size).toEqual(startSize);
    compareArrayAndSet(items, nset);
    let removedItems = 0;
    for (let i = 0; i < 100; i++) {
      const index = randInt(0, items.length - 1);
      const val = items[index];
      const valString = JSON.stringify(val);
      items = items.filter(it => { // remove all occurences from array
        if (it.length !== val.length || it[0] !== val[0]) {
          return true;
        }
        return JSON.stringify(it) !== valString;
      });
      if (nset.delete(val)) removedItems++;
    }
    expect(nset.size).toEqual(startSize - removedItems);
    compareArrayAndSet(items, nset);
  }
});


test/Fuzz.test.ts
import NSet from '../lib/NSet';

const NMAX = 50;

function randInt (min, max) {
  return Math.floor(Math.random() * (max - min + 1)) + min;
}

function compareArrayAndSet (arr, nset) {
  arr.forEach(it => {
    expect(nset.has(it)).toEqual(true);
    expect(nset.has([...it, 100])).toBe(false);
    expect(nset.has([...it.slice(0, -1), 51])).toBe(false);
  });
  const strElements = arr.map(f => JSON.stringify(f));
  nset.forEach(it => {
    const strArr = JSON.stringify(it);
    const res = strElements.find(f => f === strArr);
    expect(res).not.toBeUndefined();
  });
}

test('set-simple', () => {
  for (let n = 0; n < NMAX; n++) {
    const set = new Set();
    const nset = new NSet();
    for (let i = 0; i < 1000; i++) {
      const rnd = randInt(0, 100);
      expect(nset.has(rnd)).toBe(set.has(rnd));
      set.add(rnd);
      nset.add(rnd);
      expect(nset.has(rnd)).toBe(true);
    }
    expect(nset.size).toBe(set.size);
    for (let i = 0; i < 100; i++) {
      const rnd = randInt(0, 100);
      expect(nset.has(rnd)).toBe(set.has(rnd));
      expect(nset.delete(rnd)).toBe(set.delete(rnd));
      expect(nset.has(rnd)).toBe(false);
    }
    expect(nset.size).toBe(set.size);
    for (let i = 0; i <= 100; i++) {
      expect(nset.has(i)).toBe(set.has(i));
    }
  }
});

test('set-deep', () => {
  for (let n = 0; n < NMAX; n++) {
    let items = [];
    const nset = new NSet();
    for (let i = 0; i < 1000; i++) {
      const arr = [];
      const c = randInt(1, 9);
      for (let j = 0; j < c; j++) {
        arr.push(randInt(1, 50));
      }
      nset.add(arr);
      expect(nset.has(arr)).toBe(true);
      items.push(arr);
    }
    expect(nset.size).toBeLessThanOrEqual(1000);
    compareArrayAndSet(items, nset);
    const startSize = nset.size;
    for (let i = 0; i < 10; i++) {
      const arr = [];
      const c = randInt(1, 9);
      for (let j = 0; j < c; j++) {
        arr.push(randInt(1, 50));
      }
      if (nset.has(arr)) {
        continue;
      }
      expect(nset.delete(arr)).toBe(false);
    }
    expect(nset.size).toEqual(startSize);
    compareArrayAndSet(items, nset);
    let removedItems = 0;
    for (let i = 0; i < 100; i++) {
      const index = randInt(0, items.length - 1);
      const val = items[index];
      const valString = JSON.stringify(val);
      items = items.filter(it => { // remove all occurences from array
        if (it.length !== val.length || it[0] !== val[0]) {
          return true;
        }
        return JSON.stringify(it) !== valString;
      });
      if (nset.delete(val)) removedItems++;
    }
    expect(nset.size).toEqual(startSize - removedItems);
    compareArrayAndSet(items, nset);
  }
});


test/Object.test.ts
import NSet from '../lib/NSet';

test('add', () => {
  const set = new NSet();
  expect(set.size).toBe(0);
  
  const obj1 = { x: 5 };
  set.add(obj1);
  expect(set.size).toBe(1);
  expect(set.has(obj1)).toBe(true);
  expect(set.toJSON()).toEqual([ [ obj1 ] ]);

  set.add(obj1);
  expect(set.size).toBe(1);
  expect(set.has(obj1)).toBe(true);
  expect(set.toJSON()).toEqual([ [ obj1 ] ]);

  const obj2 = { y: 6 };
  expect(set.has(obj2)).toBe(false);
  set.add(obj2);
  expect(set.size).toBe(2);
  expect(set.has(obj2)).toBe(true);
  expect(set.toJSON()).toEqual([ [ obj1 ], [ obj2 ] ]);

  const obj3 = { y: 6, x: 10 };
  set.add(obj3);
  expect(set.size).toBe(3);
  expect(set.has(obj3)).toBe(true);
  expect(set.toJSON()).toEqual([ [ obj1 ], [ obj2 ], [ obj3 ] ]);

  set.add(obj1, obj2, obj3);
  expect(set.size).toBe(4);
  expect(set.has(obj1, obj2, obj3)).toBe(true);
  expect(set.has(obj1, obj2)).toBe(false);
  expect(set.toJSON()).toEqual([ [ obj1 ], [ obj1, obj2, obj3 ], [ obj2 ], [ obj3 ] ]);

  set.add(1, 2, 3);
  expect(set.size).toBe(5);
  expect(set.has(1, 2, 3)).toBe(true);
  expect(set.has(1)).toBe(false);
  expect(set.toJSON()).toEqual([ [ obj1 ], [ obj1, obj2, obj3 ], [ obj2 ], [ obj3 ], [1, 2, 3] ]);

  const obj4 = {};
  set.add(obj4);
  set.add(obj4);
  expect(set.size).toBe(6);
  expect(set.has(obj4)).toBe(true);
  expect(set.toJSON()).toEqual([ [ obj1 ], [ obj1, obj2, obj3 ], [ obj2 ], [ obj3 ], [1, 2, 3], [ obj4 ] ]);
});

test('clone', () => {
  const set = new NSet();
  const obj1 = { x: 5 };
  const obj2 = { y: 6 };
  const obj3 = { y: 6, x: 10 };
  set.add(obj1);
  set.add(obj2);
  set.add(obj3);
  set.add(obj1, obj2, obj3);

  const set2 = set.clone();
  expect(set2.toJSON()).toEqual([ [ obj1 ], [ obj1, obj2, obj3 ], [ obj2 ], [ obj3 ] ]);
});

test('delete', () => {
  const set = new NSet();
  const obj1 = { x: 5 };
  const obj2 = { y: 6 };
  const obj3 = { y: 6, x: 10 };
  set.add(obj1);
  set.add(obj2);
  set.add(obj3);
  set.add(obj1, obj2, obj3);

  expect(set.delete({})).toBe(false);
  expect(set.delete({ x: 5 })).toBe(false);
  expect(set.size).toBe(4);
  expect(set.delete(obj3)).toBe(true);
  expect(set.size).toBe(3);
  expect(set.delete(obj1, obj2)).toBe(false);
  expect(set.delete(obj1, obj2, obj3)).toBe(true);
  expect(set.size).toBe(2);
  expect(set.toJSON()).toEqual([ [ obj1 ], [ obj2 ] ]);
});



lib/NSet.ts
interface RecursiveNode<T> extends Map<T, Set<T> | RecursiveNode<T>> {}

type Node<T> = Set<T> | RecursiveNode<T>;

export default class NSet<T> {
  private _data: Node<T>;
  private _size: number;

  constructor (arr?: T[] | NSet<T> | Set<T>) {
    this._size = 0;
    if (arr === undefined) {
      this._data = new Set<T>();
      return;
    }

    if (arr instanceof NSet) {
      const newSet = arr.clone();
      this._data = newSet._data;
      this._size = newSet._size;
      return;
    }

    if (arr instanceof Set) {
      this._size = arr.size;
      this._data = new Set<T>(arr);
      return;
    }

    if (Array.isArray(arr)) {
      this._data = new Set<T>();
      arr.forEach(item => {
        this.add(item);
      });
      return;
    }
    throw new Error('Invalid type for NSet constructor');
  }

  get size (): number {
    return this._size;
  }

  clear (): void {
    this._data = new Set<T>();
    this._size = 0;
  }

  private _cloneWalk (node: Node<T>): Node<T> {
    if (node instanceof Set) {
      return new Set(node);
    }
    const map = new Map();
    node.forEach((value, key) => {
      map.set(key, !value ? value : this._cloneWalk(value));
    });
    return map;
  }

  clone (): NSet<T> {
    const newSet = new NSet<T>();
    newSet._size = this._size;
    newSet._data = this._cloneWalk(this._data);
    return newSet;
  }

  private _addWalk (arr: T[]): Node<T> {
    const forLength = arr.length - 1;
    let posParent = null;
    let pos = this._data;
    let item = arr[0];
    for (let i = 0; i < forLength; i++) {
      if (item === undefined) {
        throw new TypeError('Undefined values are unsupported by NSet');
      }
      if (pos instanceof Set) {
        const map: RecursiveNode<T> = new Map();
        pos.forEach(it => map.set(it, null));
        if (map.has(item)) {
          map.set(item, new Set<T>([undefined]));
        } else {
          map.set(item, new Set<T>());
        }
        if (posParent === null) {
          this._data = map;
        } else {
          posParent.set(arr[i - 1], map);
        }
        pos = map;
      } else if (!pos.has(item)) { // already is a Map
        if (i < arr.length - 2) {
          pos.set(item, new Map());
        } else {
          pos.set(item, new Set<T>());
        }
      } else if (pos.get(item) === null) {
        pos.set(item, new Set<T>([undefined]));
      }
      posParent = pos;
      pos = pos.get(item);
      item = arr[i + 1];
    }
    return pos;
  }

  add (value: T[]): NSet<T>;
  add (...values: T[]): NSet<T>;
  add (...values): NSet<T> {
    const arr = Array.isArray(values[0]) ? values[0] as unknown as T[] : values as T[];
    return this._add(arr);
  }

  private _add (arr: T[]): NSet<T> {
    const pos = arr.length > 1 ? this._addWalk(arr) : this._data;
    const item = arr[arr.length - 1];
    if (item === undefined) {
      throw new TypeError('Undefined values are unsupported by NSet');
    }
    // processing last index
    if (pos instanceof Set) {
      const s1 = pos.size;
      pos.add(item);
      this._size += pos.size - s1;
    } else {
      const it = pos.get(item);
      if (it === undefined) {
        pos.set(item, null);
        this._size++;
      } else if (it === null) {
      } else if (it instanceof Map) {
        const s1 = it.size;
        it.set(undefined, null);
        this._size += it.size - s1;
      } else { // it is Set
        const s1 = it.size;
        it.add(undefined);
        this._size += it.size - s1;
      }
    }
    return this;
  }

  private _walkToItem (arr: T[]): false | Node<T> {
    const forLength = arr.length - 1;
    let pos = this._data;
    let item = arr[0];
    for (let i = 0; i < forLength; i++) {
      if (item === undefined) {
        throw new TypeError('Undefined values are unsupported by NSet');
      }
      if (!(pos instanceof Map)) {
        return false;
      }
      pos = pos.get(item);
      if (pos === undefined) {
        return false;
      }
      item = arr[i + 1];
    }
    return pos;
  }

  has (value: T[]): boolean;
  has (...values: T[]): boolean;
  has (...values): boolean {
    const arr = Array.isArray(values[0]) ? values[0] as unknown as T[] : values as T[];
    return this._has(arr);
  }

  private _has (arr: T[]): boolean {
    let pos = this._data;
    if (arr.length > 1) {
      const res = this._walkToItem(arr);
      if (res === false) {
        return false;
      }
      pos = res;
    }
    const item = arr[arr.length - 1];

    if (item === undefined) {
      throw new TypeError('Undefined values are unsupported by NSet');
    }

    if (pos === null) {
      return false;
    }

    if (pos instanceof Set) {
      return pos.has(item);
    } else { // Map
      const it = pos.get(item);
      if (it === undefined) {
        return false;
      }
      if (it === null) {
        return true;
      }
      if (it.has(undefined)) {
        return true;
      }
    }
    return false; // pos.has(undefined);
  }

  delete (value: T[]): boolean;
  delete (...values: T[]): boolean;
  delete (...values): boolean {
    const arr = Array.isArray(values[0]) ? values[0] as unknown as T[] : values as T[];
    return this._delete(arr);
  }

  private _delete (arr: T[]): boolean {
    let pos = this._data;
    if (arr.length > 1) {
      const res = this._walkToItem(arr);
      if (res === false) {
        return false;
      }
      pos = res;
    }
    const item = arr[arr.length - 1];
    if (item === undefined) {
      throw new TypeError('Undefined values are unsupported by NSet');
    }

    if (pos instanceof Set) {
      const res = pos.delete(item);
      if (res) this._size--;
      return res;
    } else if (pos instanceof Map) {
      const it = pos.get(item);
      if (it === undefined) {
        return false;
      }
      if (it === null) {
        pos.delete(item);
        this._size--;
        return true;
      }
      // Has Map or Set
      const res = it.delete(undefined);
      if (res) this._size--;
      return res;
    }
    return false;
  }

  entries (): Iterator<[T[], T[]]> {
    const values = this.values();
    return {
      next: () => {
        const it = values.next();
        return {
          value: [it.value, it.value],
          done: it.done,
        };
      }
    };
  }

  values (): Iterator<T[]> {
    const iterators = [this._data.entries()];
    const values: T[] = [];
    let iteratorIndex: number = 0;
    let inSet: boolean = this._data instanceof Set;

    return {
      next: () => {
        while (true) {
          const it = iterators[iteratorIndex];
          const v = it.next();

          if (iteratorIndex === 0 && v.done) {
            break;
          } else if (v.done) {
            inSet = false;
            values[iteratorIndex] = undefined;
            iteratorIndex--;
            continue;
          }
          values[iteratorIndex] = v.value[0];
          const content = v.value[1];
          if (v.value[0] === undefined) { // terminator
            const newItem = values.slice(0, iteratorIndex);
            return {
              value: newItem,
              done: false
            };
          } else if (content === null || inSet) {
            const newItem = values.slice(0, iteratorIndex + 1);
            return {
              value: newItem,
              done: false
            };
          } else if ((content as Node<T>).size > 0) {
            iteratorIndex++;
            inSet = content instanceof Set;
            iterators[iteratorIndex] = (content as Node<T>).entries();
          }
        }
        return {
          value: undefined,
          done: true
        };
      }
    };
  }

  forEach (func: (value: T[], index: number, set: NSet<T>) => void): void {
    const entries = this.values();
    let index = 0;
    while (true) {
      const v = entries.next();
      if (v.done) {
        break;
      }
      func(v.value, index++, this);
    }
  }

  toJSON (): T[][] {
    const obj: T[][] = [];
    const entries = this.values();
    while (true) {
      const v = entries.next();
      if (v.done) {
        break;
      }
      obj.push(v.value);
    }
    return obj;
  }

  [Symbol.iterator] (): Iterator<T[]> {
    return this.values();
  }

  inspect (): string {
    const entries = this.values();
    const pairs = [];
    while (true) {
      const v = entries.next();
      if (v.done) {
        break;
      }
      pairs.push(v.value);
    }

    return `NSet { ${pairs.map(p => `(${p.join(', ')})`).join(', ')} }`;
  }
}


benchmark.js
const N = 5;
const NS_PER_SEC = 1e9;

const NSet = require('./');

function randInt (min, max) {
  return Math.floor(Math.random() * (max - min + 1)) + min;
}

const setItems = [];
for (let i = 0; i < 1e6; i++) {
  const it = [];
  const len = randInt(1, 4);
  for (let j = 0; j < len; j++) {
    it.push(randInt(0, 999));
  }
  setItems.push(it);
}

console.log(`Data distribution:\n------------`);
console.log(`= 1-depth - ${setItems.filter(it => it.length === 1).length}`);
console.log(`= 2-depth - ${setItems.filter(it => it.length === 2).length}`);
console.log(`= 3-depth - ${setItems.filter(it => it.length === 3).length}`);
console.log(`= 4-depth - ${setItems.filter(it => it.length === 4).length}\n\n`);

const tests = [];

const setItems1 = setItems.filter(it => it.length === 1);

tests.push(() => {
  const map = new Map();
  for (let i = 0; i < setItems1.length; i++) {
    map.set(setItems1[i][0], true);
  }
  for (let i = 0; i < setItems1.length; i++) {
    if (!map.has(setItems1[i][0])) {
      return false;
    }
  }
  for (let i = setItems1.length - 1; i >= 0; i--) {
    map.delete(setItems1[i][0]);
  }
  return {name: 'Map', value: map.size};
});

tests.push(() => {
  const set = new Set();
  for (let i = 0; i < setItems1.length; i++) {
    set.add(setItems1[i][0]);
  }
  for (let i = 0; i < setItems1.length; i++) {
    if (!set.has(setItems1[i][0])) {
      return false;
    }
  }
  for (let i = setItems1.length - 1; i >= 0; i--) {
    set.delete(setItems1[i][0]);
  }
  return {name: 'Set', value: set.size};
});

tests.push(() => {
  const set = new NSet();
  for (let i = 0; i < setItems1.length; i++) {
    set.add(setItems1[i]);
  }
  for (let i = 0; i < setItems1.length; i++) {
    if (!set.has(setItems1[i])) {
      return false;
    }
  }
  for (let i = setItems1.length - 1; i >= 0; i--) {
    set.delete(setItems1[i]);
  }
  return {name: 'NSet 1-depth', value: set.size};
});

const setItems2 = setItems.filter(it => it.length === 2);

tests.push(() => {
  const set = new NSet();

  for (let i = 0; i < setItems2.length; i++) {
    set.add(...setItems2[i]);
  }
  for (let i = 0; i < setItems2.length; i++) {
    if (!set.has(...setItems2[i])) {
      return false;
    }
  }
  for (let i = setItems2.length - 1; i >= 0; i--) {
    set.delete(...setItems2[i]);
  }
  return {name: 'NSet 2-depth', value: set.size};
});

const setItems3 = setItems.filter(it => it.length === 3);

tests.push(() => {
  const set = new NSet();

  for (let i = 0; i < setItems3.length; i++) {
    set.add(...setItems3[i]);
  }
  for (let i = 0; i < setItems3.length; i++) {
    if (!set.has(...setItems3[i])) {
      return false;
    }
  }
  for (let i = setItems3.length - 1; i >= 0; i--) {
    set.delete(...setItems3[i]);
  }
  return {name: 'NSet 3-depth', value: set.size};
});

tests.push(() => {
  const set = new NSet();

  for (let i = 0; i < setItems.length; i++) {
    set.add(...setItems[i]);
  }
  for (let i = 0; i < setItems.length; i++) {
    if (!set.has(...setItems[i])) {
      return false;
    }
  }
  for (let i = setItems.length - 1; i >= 0; i--) {
    set.delete(...setItems[i]);
  }
  return {name: 'NSet 4-depth', value: set.size};
});

const results = [];

for (let n = 0; n < N; n++) {
  tests.forEach((test, key) => {
    if (!results[key]) {
      results[key] = {
        results: [],
        timings: []
      };
    }
    const startTime = process.hrtime();
    const res = tests[key](); // run test
    const duration = process.hrtime(startTime);
    results[key].results.push(res);
    results[key].timings.push((duration[0] * NS_PER_SEC + duration[1]) / 1000000);
  });
}

function formatTiming (ms) {
  return `${ms.toFixed(3)} ms`;
}

console.log(`After ${N} runs:\n------------`);
results.forEach(result => {
  result.timings.sort();
  const sum = result.timings.reduce((sum, x) => sum + x);
  const avg = sum / N;
  const middle = Math.floor(N / 2);
  const median = N % 2 === 0 ? (result.timings[middle] + result.timings[middle - 1]) / 2 : result.timings[middle];
  const min = Math.min(...result.timings);
  const max = Math.max(...result.timings);
  console.log(`${result.results[0].name.padEnd(20)} - Min: ${formatTiming(min)}, Max: ${formatTiming(max)}, Mean: ${formatTiming(avg)}, Median: ${formatTiming(median)}`);
});


clone_benchmark.ts
import NSet from './lib/NSet';

function randInt (min, max) {
  return Math.floor(Math.random() * (max - min + 1)) + min;
}

const setItems = [];
for (let i = 0; i < 1e6; i++) {
  const it = [];
  const len = randInt(1, 4);
  for (let j = 0; j < len; j++) {
    it.push(randInt(0, 999));
  }
  setItems.push(it);
}

const set = new NSet(setItems);
// warmup
const s1 = new NSet();
set.forEach(value => s1.add(value));
const s2 = set.clone();


for (let i = 0; i < 5; i++) {
  console.time('clone forEach');
  const clonedWithForEach = new NSet();
  set.forEach(value => clonedWithForEach.add(value));
  console.log(clonedWithForEach.size);
  console.timeEnd('clone forEach');
  console.time('clone func');
  const clonedWithFunc = set.clone();
  console.log(clonedWithFunc.size);
  console.timeEnd('clone func');
}

benchmark.ts
import NSet from './lib/NSet';

const N = 5;
const NS_PER_SEC = 1e9;

function randInt (min, max) {
  return Math.floor(Math.random() * (max - min + 1)) + min;
}

const setItems = [];
for (let i = 0; i < 1e6; i++) {
  const it = [];
  const len = randInt(1, 4);
  for (let j = 0; j < len; j++) {
    it.push(randInt(0, 999));
  }
  setItems.push(it);
}

console.log(`Data distribution:\n------------`);
console.log(`= 1-depth - ${setItems.filter(it => it.length === 1).length}`);
console.log(`= 2-depth - ${setItems.filter(it => it.length === 2).length}`);
console.log(`= 3-depth - ${setItems.filter(it => it.length === 3).length}`);
console.log(`= 4-depth - ${setItems.filter(it => it.length === 4).length}\n\n`);

const tests = [];

const setItems1 = setItems.filter(it => it.length === 1);

tests.push(() => {
  const map = new Map();
  for (let i = 0; i < setItems1.length; i++) {
    map.set(setItems1[i][0], true);
  }
  for (let i = 0; i < setItems1.length; i++) {
    if (!map.has(setItems1[i][0])) {
      return false;
    }
  }
  for (let i = setItems1.length - 1; i >= 0; i--) {
    map.delete(setItems1[i][0]);
  }
  return {name: 'Map', value: map.size};
});

tests.push(() => {
  const set = new Set();
  for (let i = 0; i < setItems1.length; i++) {
    set.add(setItems1[i][0]);
  }
  for (let i = 0; i < setItems1.length; i++) {
    if (!set.has(setItems1[i][0])) {
      return false;
    }
  }
  for (let i = setItems1.length - 1; i >= 0; i--) {
    set.delete(setItems1[i][0]);
  }
  return {name: 'Set', value: set.size};
});

tests.push(() => {
  const set = new NSet();
  for (let i = 0; i < setItems1.length; i++) {
    set.add(setItems1[i]);
  }
  for (let i = 0; i < setItems1.length; i++) {
    if (!set.has(setItems1[i])) {
      return false;
    }
  }
  for (let i = setItems1.length - 1; i >= 0; i--) {
    set.delete(setItems1[i]);
  }
  return {name: 'NSet 1-depth', value: set.size};
});

const setItems2 = setItems.filter(it => it.length === 2);

tests.push(() => {
  const set = new NSet();

  for (let i = 0; i < setItems2.length; i++) {
    set.add(...setItems2[i]);
  }
  for (let i = 0; i < setItems2.length; i++) {
    if (!set.has(...setItems2[i])) {
      return false;
    }
  }
  for (let i = setItems2.length - 1; i >= 0; i--) {
    set.delete(...setItems2[i]);
  }
  return {name: 'NSet 2-depth', value: set.size};
});

const setItems3 = setItems.filter(it => it.length === 3);

tests.push(() => {
  const set = new NSet();

  for (let i = 0; i < setItems3.length; i++) {
    set.add(...setItems3[i]);
  }
  for (let i = 0; i < setItems3.length; i++) {
    if (!set.has(...setItems3[i])) {
      return false;
    }
  }
  for (let i = setItems3.length - 1; i >= 0; i--) {
    set.delete(...setItems3[i]);
  }
  return {name: 'NSet 3-depth', value: set.size};
});

tests.push(() => {
  const set = new NSet();

  for (let i = 0; i < setItems.length; i++) {
    set.add(...setItems[i]);
  }
  for (let i = 0; i < setItems.length; i++) {
    if (!set.has(...setItems[i])) {
      return false;
    }
  }
  for (let i = setItems.length - 1; i >= 0; i--) {
    set.delete(...setItems[i]);
  }
  return {name: 'NSet all-depth', value: set.size};
});

const results = [];

for (let n = 0; n < N; n++) {
  tests.forEach((test, key) => {
    if (!results[key]) {
      results[key] = {
        results: [],
        timings: []
      };
    }
    const startTime = process.hrtime();
    const res = tests[key](); // run test
    const duration = process.hrtime(startTime);
    results[key].results.push(res);
    results[key].timings.push((duration[0] * NS_PER_SEC + duration[1]) / 1000000);
  });
}

function formatTiming (ms) {
  return `${ms.toFixed(3)} ms`;
}

console.log(`After ${N} runs:\n------------`);
results.forEach(result => {
  result.timings.sort();
  const sum = result.timings.reduce((sum, x) => sum + x);
  const avg = sum / N;
  const middle = Math.floor(N / 2);
  const median = N % 2 === 0 ? (result.timings[middle] + result.timings[middle - 1]) / 2 : result.timings[middle];
  const min = Math.min(...result.timings);
  const max = Math.max(...result.timings);
  console.log(`${result.results[0].name.padEnd(20)} - Min: ${formatTiming(min)}, Max: ${formatTiming(max)}, Mean: ${formatTiming(avg)}, Median: ${formatTiming(median)}`);
});
